### ISSUE TITLE
Take another look at learning rate schedulers

### ISSUE THREAD

**Issue by samedii** [codewell/ml-workflow#46]
Is there a better way of doing this? It should just be a mathematical function of the number of iterations.

**Comment by FelixAbrahamsson** [comment]
Right now the wrappers we have look so thin that they feel redundant. I agree I would also like to have a better way to do this, or just skip having the LR schedulers in the workflow package entirely and just keep them in a template.

**Comment by samedii** [comment]
Let's fix and discuss this after the utter challenge is over. I think the solution seems extremely over-engineered

**Comment by FelixAbrahamsson** [comment]
Ignite has `ignite.contrib.handlers.param_scheduler.LRScheduler(lr_scheduler, save_history=False, **kwds)` where `lr_scheduler` is an optimizer from `torch.optim.lr_scheduler`. Torch has `torch.optim.lr_scheduler.LambdaLR(optimizer, lr_lambda, last_epoch=-1)` where `lr_lambda` can be a function that takes an iteration count as input and outputs a learning rate.
I think we can then do something like
```python
def get_cyclical_lr(cycle_size, min, max):
    def cyclical_lr(step):
        return (1 - (step % cycle_size) / cycle_size) * (max - min) + min
    return cyclical_lr


def get_decaying_lr(decay):
    def decay_lr(lr, step):
        return lr * decay ** step
    return decay_lr


lr_schedule_fn = starcompose(
    lambda step: (get_cyclical_lr(100, 0.01, 0.1)(step), step),
    get_decaying_lr(.99),
)


from ignite.contrib.handlers.param_scheduler import LRScheduler
from torch.optim.lr_scheduler import LambdaLR

lr_scheduler = LRScheduler(
    LambdaLR(optimizer, lr_schedule_fn)
)
trainer.add_event_handler(
    Events.ITERATION_STARTED, 
    lr_scheduler
)
```
This would give the following lr scheduling
![image](https://user-images.githubusercontent.com/19688421/74048591-d09aff00-49d2-11ea-92cd-6807a404b12b.png)

**Comment by samedii** [comment]
We should decide on a common prefix for when we construct a function? I normally use `create_`, in coop we have for example: `create_preprocess`

@Jim-Holmstroem said that in rust they use `mk_` but that does not feel so pythonic. Maybe `make_ `?

Note that when these functions are used as decorators then we don't have any prefix though...

This looks pretty ok too?
```
@warmup(n_steps=100)
def my_learning_rate(step):
    pass
```

**Comment by samedii** [comment]
We should consider creating "flat" functions as well and letting the user use partial to set an argument maybe?

I do not really like this implementation but I put some time into it so I'll post it anyway :)
```
lr_scheduler = LRScheduler(
    LambdaLR(
        optimizer,
        starcompose(
            lambda step: step, config['learning_rate'],
            partial(warmup, n_steps=100),
            partial(cyclical, length=50, relative_minimum=0.1),
            lambda step, lr: lr,
        )
    )
)

```

```
def exponential_decay(step, lr, decay=0.97):
    return step, lr * decay ** step
```

I like the idea of using a relative minimum instead
```
def cyclical(step, lr, length=100, relative_min=0.1):
    return step, lr * (1 - (step % length) / length) * (1- relative_min) + relative_min
```

**Comment by samedii** [comment]
Brain explosion? :P

    @trainer.on(Events.ITERATION_STARTED)
    @LRScheduler
    @partial(LambdaLR, optimizer)
    def learning_rate(step):
        step = warmup(n_steps=100)
        return cyclical(
            step, config['learning_rate'], length=50, relative_minimum=0.1
        )

**Comment by samedii** [comment]
I think something like this that abstracts away that LambdaLR wants a function of only step while we have a function of two variables (step and learning_rate).

```
LearningRateScheduler(
        optimizer,
        config['learning_rate'],
        starcompose(
            partial(warmup, n_steps=100),
            partial(cyclical, length=50, relative_minimum=0.1),
        )
    )
).attach(trainer)
```

Do you prefer skipping partial for a slightly more difficult library but slightly easier usage?
```
LearningRateScheduler(
        optimizer,
        config['learning_rate'],
        starcompose(
            warmup(n_steps=100),
            cyclical(length=50, relative_minimum=0.1),
        )
    )
).attach(trainer)
```

**Comment by Jim-Holmstroem** [comment]
another possibility is to have
```
def DecayingLR(decay):
    def decay_lr(lr, step):
        return lr * decay ** step
    return decay_lr
```
feels kinda natural when you use it
```
starcompose(
    DecayingLR(0.3),
    CyclicalLR(100, 1, 20),
) # non-sense just to prove readability
```
but in the end, we'll use something that everyone is comfortable with, personally I think a lot of the
non-descriptive prefixes etc can be removed from naming when possible
like `get_` `create_` and instead be used when something is actually "physically" retrieved or created (like database rows or creating threads)

**Comment by samedii** [comment]
> another possibility is to have
> 
> ```
> def DecayingLR(decay):
>     def decay_lr(lr, step):
>         return lr * decay ** step
>     return decay_lr
> ```
I think we should only let classes have CamelCase.

I agree with the idea of having functions of two variables. That part is the same as I am proposing above. (We need that wrapping function that I talked about to do this nicely though because torch expects a function of only step)

I also think not having get_ make_ create_ as a prefix made it nice to read.

**Comment by Jim-Holmstroem** [comment]
There is nothing special about a class, you are actually in this case just binding `decay` to the closure of `decay_lr`
Seen from the outside it's functionally equivalent to (the much more pythonic)
```
class DecayingLR:
    def __init__(self, decay):
        self.decay = decay
    def __call__(self, lr, step):
        return lr * self.decay ** step
```
If something it would be much more important to show that this "thing" is impure and has side-effects compared to that it's a class or a function
(on a side note on that I think you should use side-effects as little as possible and when you do it should be either "you can think of it as pure, just a technicality" or that it's blatantly obvious that it has side-effects

**Comment by samedii** [comment]
I also thought about a class with `__call__` :) If want a less advanced implementation then I think using partial is nice.

I think a class with `__call__` is okay too. I just don't want functions with CamelCase because it's against PEP that we've been following everywhere. My personal preference is that CamelCase is quite ugly and I prefer closures over classes where the internal state is easily accessible while a closure has less issues with such possible side effects

I can get bind having something to show that a function is impure if we can do it in a nice way :) In pytorch they postfix `_` if an operation is `in-place`
